src/condition/ssrf_detector.cpp

```diff
@@ -0,0 +1,222 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#include "condition/ssrf_detector.hpp"
+#include "exception.hpp"
+#include "iterator.hpp"
+#include "uri_utils.hpp"
+#include "utils.hpp"
+#include <iostream>
+
+using namespace std::literals;
+
+namespace ddwaf {
+
+namespace {
+
+constexpr std::array<std::string_view, 10> dangerous_ips{"169.254.0.0/16", "127.0.0.1/32",
+    "10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16", "100.64.0.0/10", "::1/128", "fc00::/7",
+    "fe80::/10", "2001:db8:1234:1a00::/56"};
+
+constexpr const std::array<std::string_view, 10> dangerous_domains{
+    "metadata.google", "burpcollaborator.net", ".local", ".internal", "ram.aliyuncs.com",
+    "ifconfig.pro", "localhost", "fuf.me", "localtest.me", "ulh.us" // Legacy domains
+};
+
+constexpr const std::array<std::string_view, 4> authorised_schemes{"https", "http", "ftps", "ftp"};
+
+constexpr const auto &npos = std::string_view::npos;
+
+using ssrf_result = std::optional<std::pair<std::string, std::vector<std::string>>>;
+
+bool detect_parameter_injection(
+    const uri_decomposed &uri, std::string_view param, std::size_t param_index)
+{
+    const auto path_end = uri.path_index + uri.path.size();
+
+    // REST parameter injection will involve introducing a / to the URL, however, such
+    // slashes are allowed after the ? or the #. Check if the slash is within the parameters.
+    if (auto slash_index = param.find('/'); slash_index != npos) {
```

Comment:
I find a bit odd this approach of two statement within the conditional. I would prioritise readability over not having a variable with longer scoper

Reply:
This is perfectly readable.